### ISSUE TITLE
Define bug, define that include other define can be broken at some point

### ISSUE THREAD

**Issue by danilw** [KhronosGroup/glslang#2368]
**Bug code**:
```c

#define arg_two_(arg_, arg_2, gidx_, func_) (func_(arg_, arg_2) + 0.)

#define min(arg_,arg_2) arg_two_(arg_, arg_2, 12, min)
#define max(arg_,arg_2) arg_two_(arg_, arg_2, 12, max)

//why this work?
//#define max(arg_,arg_2) arg_two_(arg_, arg_2, 12, min) 

void test(){
    float a=min(1., max(1.,1.)); //bug
    
    //float a=min(1., min(1.,1.)); //why this work?
}
```

**Error message**:
```
ERROR: main.glsl:11: 'max' : undeclared identifier 
ERROR: main.glsl:11: 'arg_two_' : no matching overloaded function found 
```

Each marked line does not have bug, bug only when `max` used as argument. 

Other GLSL compilers (even OpenGL in drivers) do not have this but, this is only in glslangValidator

```
glslangValidator --version
Glslang Version: 7.13.3496
ESSL Version: OpenGL ES GLSL 3.20 glslang Khronos. 13.3496
GLSL Version: 4.60 glslang Khronos. 13.3496
SPIR-V Version 0x00010400, Revision 1
GLSL.std.450 Version 100, Revision 1
Khronos Tool ID 8
SPIR-V Generator Version 7
GL_KHR_vulkan_glsl version 100
ARB_GL_gl_spirv version 100
```

**Comment by danilw** [comment]
Related to https://www.shadertoy.com/view/tllfDM
this shader can not be build in glslangValidator because of lots errors

I upload this shader for glslangValidator  download 
[Debug_shader.zip](https://github.com/KhronosGroup/glslang/files/5048803/Debug_shader.zip)

go to `Debug_shader/shaders` and launch `build_shaders.cmd` or `build_shaders.sh` or copy its commands to command line
shaders can not be compiled to spirv

**Comment by therontarigo** [comment]
Consider simplified macros and the following test cases:
`preproc-test.h`
```C
#define APPLY(func, arg) func(arg)

// underscore+name : do-nothing function to track macro invocations
#define M(x) mfunc(x)
#define A(x) APPLY(afunc,_a(x))
#define B(x) APPLY(M,_b(x))
#define C(x) APPLY(C,_c(x))

#define X0(x) x
#define X1(x) X0(x)
#define X2(x) X1(x)
#define X3(x) X2(x)
#define X4(x) X3(x)

// Nested APPLY of functions works as expected
case00: A(_outer(A(0)))
// Nested APPLY of macros works as expected
case01: B(_outer(B(0)))
// Nested APPLY of same-name function works as expected
case02: C(_outer(C(0)))
// Show behavior of C(0) outside of any other macro
case03: _outer(C(0))
// Show C(0) used as argument to various macros
case04: M(_outer(C(0)))
case05: A(_outer(C(0)))
case06: B(_outer(C(0)))
case07: C(_outer(C(0)))
case08: X0(_outer(C(0)))
case09: X1(_outer(C(0)))
case10: X2(_outer(C(0)))
case11: X3(_outer(C(0)))
case12: X4(_outer(C(0)))
```
Khronos OpenGL Shading Language Standard:
"There is a preprocessor that processes the source strings as part of the compilation process. Except as noted below, it behaves as the C++ standard preprocessor" (No exceptions pertaining to macro evaluation behavior follow.)
"#define  and  #undef  functionality  are  defined  as  is  standard  for  C++  preprocessors  for  macro definitions both with and without macro parameters."
So, where GCC and Clang appear to implement well-tested standard-compliant C++ preprocessors (note there is no difference from C preprocessor here), the expected behavior follows:
`cc -x c++ -E preproc-test.h` ->
```
case00: afunc(_a(_outer(afunc(_a(0)))))
case01: mfunc(_b(_outer(mfunc(_b(0)))))
case02: C(_c(_outer(C(_c(0)))))
case03: _outer(C(_c(0)))
case04: mfunc(_outer(C(_c(0))))
case05: afunc(_a(_outer(C(_c(0)))))
case06: mfunc(_b(_outer(C(_c(0)))))
case07: C(_c(_outer(C(_c(0)))))
case08: _outer(C(_c(0)))
case09: _outer(C(_c(0)))
case10: _outer(C(_c(0)))
case11: _outer(C(_c(0)))
case12: _outer(C(_c(0)))
```
whereas
`glslangValidator -S frag -E preproc-test.h` ->
```
case00 : afunc(_a(_outer(afunc(_a(0)))))
case01 : mfunc(_b(_outer(mfunc(_b(0)))))
case02 : C(_c(_outer(C(_c(0)))))
case03 : _outer(C(_c(0)))
case04 : mfunc(_outer(C(_c(_c(0)))))
case05 : afunc(_a(_outer(APPLY(C, _c(_c(_c(0)))))))
case06 : mfunc(_b(_outer(APPLY(C, _c(_c(_c(0)))))))
case07 : C(_c(_outer(C(_c(0)))))
case08 : _outer(C(_c(_c(0))))
case09 : _outer(C(_c(_c(_c(0)))))
case10 : _outer(C(_c(_c(_c(_c(0))))))
case11 : _outer(C(_c(_c(_c(_c(_c(0)))))))
case12 : _outer(C(_c(_c(_c(_c(_c(_c(0))))))))
```

Through a test shader it can be seen that Mesa and WebGL2 (Firefox) follow the standard behavior, while Nvidia follows glslangValidator behavior.

Conclusion: Standard C/C++ and OpenGL preprocessors stop expanding macros as soon as recursion is encountered.  The second `C` in `#define C(x) APPLY(C,_c(x))` is never interpreted as a the name of a macro.

glslangValidator and Nvidia however re-interpret `C(x)` as a macro invocation each time it appears in the (expanded) argument to another macro.

**Comment by therontarigo** [comment]
That observation doesn't leave the appearance of APPLY in the output for cases 05,06 fully explained.  It looks like some sort of recursion failsafe, since
```C
#define APPLY0(func, arg) func(arg)
#define APPLY1(func, arg) func(arg)

// underscore+name : do-nothing function to track macro invocations
#define M(x) mfunc(x)
#define A(x) APPLY1(afunc,_a(x))
#define B(x) APPLY1(M,_b(x))
#define C(x) APPLY0(C,_c(x))

case05: A(_outer(C(0)))
case06: B(_outer(C(0)))
```
becomes fully expanded, although still with the spurious applications of `C(x)`

**Comment by danilw** [comment]
so you mean this my case is Undefined Behavior in defines?

**Comment by therontarigo** [comment]
> so you mean this my case is Undefined Behavior in defines?
No, I think glslangValidator is behaving incorrectly for a reasonable case, it is a bug.
> C/C++ defines should "unroll" define loops one by one without applying full "define loop" again when define in define found...
Right, that's the expected behavior, I don't think what glslangValidator is doing was intended.

**Comment by arcady-lunarg** [comment]
This was fixed by #3114. Closing.
